src/NHibernate/Engine/Query/NativeSQLQueryPlan.cs

```diff
@@ -151,19 +151,18 @@ private SqlString ExpandDynamicFilterParameters(SqlString sqlString, ICollection
 						string parameterName = parts[1];
 						var filter = (FilterImpl)enabledFilters[filterName];
 
-						object value = filter.GetParameter(parameterName);
+						var listValue = filter.GetParameterList(parameterName);
 						IType type = filter.FilterDefinition.GetParameterType(parameterName);
 						int parameterColumnSpan = type.GetColumnSpan(session.Factory);
-						var collectionValue = value as ICollection;
 						int? collectionSpan = null;
 
 						// Add query chunk
 						string typeBindFragment = string.Join(", ", Enumerable.Repeat("?", parameterColumnSpan).ToArray());
 						string bindFragment;
-						if (collectionValue != null && !type.ReturnedClass.IsArray)
+						if (listValue != null && !type.ReturnedClass.IsArray)
 						{
-							collectionSpan = collectionValue.Count;
-							bindFragment = string.Join(", ", Enumerable.Repeat(typeBindFragment, collectionValue.Count).ToArray());
+							collectionSpan = listValue.Cast<object>().Count();
```

Comment:
Wouldn't it result in full enumeration and unnecessary boxing for all struct elements?
If we still need to deal with IEnumerable interface I think it's better to add internal Count() extension:
```C#
internal static int Count(this IEnumerable source)
{
	if (source is ICollection col)
		return col.Count;

	int count = 0;
	var e = source.GetEnumerator();
	//in case it's generic disposable enumerator
	using (e as IDisposable)
	{
		while (e.MoveNext())
		{
			count++;
		}
	}
	return count;
}
```

Reply:
Yes you are right. But I was unwilling to add this to the recently obsoleted `EnumerableExtensions` utility.

Reply:
It's obsoleted to be public. I think it's still can survive as internal helper.

Reply:
Commit amended.

Reply:
>Wouldn't it result in full enumeration and unnecessary boxing for all struct elements?

No it would not.

Reply:
>Not in every case.
Could you give an example? I'm looking at Cast and Count implementation and don't see any shortcuts.

Reply:
>Could you give an example?

#### Cast

`IEnumerable<T>` is co-variant. `Cast` has [a shortcut for it](https://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,1023):
```csharp
public static IEnumerable<TResult> Cast<TResult>(this IEnumerable source) {
  IEnumerable<TResult> typedSource = source as IEnumerable<TResult>;
  if (typedSource != null) return typedSource;
  if (source == null) throw Error.ArgumentNull("source");
  return CastIterator<TResult>(source);
}
```
If collection is generic, cast will chose this path. `CastIterator` would be involved only if collection is not generic.

In our case we first cast `ICollection<T>` to `IEnumerable` then do `.Cast<T>` which will do the upcast to `IEnumerable<T>` without involving `CastIterator`.

#### Count

Count [has shortcuts also](https://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,1305):

```
public static int Count<TSource>(this IEnumerable<TSource> source) {
  if (source == null) throw Error.ArgumentNull("source");
  ICollection<TSource> collectionoft = source as ICollection<TSource>;
  if (collectionoft != null) return collectionoft.Count;
  ICollection collection = source as ICollection;
  if (collection != null) return collection.Count;
  int count = 0;
  using (IEnumerator<TSource> e = source.GetEnumerator()) {
    checked {
      while (e.MoveNext()) count++;
    }
  }
  return count;
}
```

Because we did not perform `CastIterator` it will just call `.Count` on our collection.

Reply:
It seems I misread your comment. Indeed for *structs* it will always result in a cast.